posts/kotlin-like-java-but-better-1.md

```diff
@@ -0,0 +1,180 @@
+Semicolons are optional in Kotlin, and you shouldn't ever need them. The 
+compiler will infer your statement endings.
+
+## Null Safety
+One of the key design decisions that the Kotlin team made was to enforce 
+[null safety](https://kotlinlang.org/docs/null-safety.html) at compile-time.
+All variables must explicitly be declared as either nullable or non-nullable. 
+```
+var foo: Int? // nullable Int variable
+var bar: Int // non-nullable Int
+
+bar = Null // compilation error
+```
+As a result of this feature, NullPointerExceptions will never be thrown in 
+pure Kotlin code unless you explicitly throw one. They won't be missed! Kotlin
+also provides some additional syntax to enable working with these new null 
+constraints. 
+
+Inline null-checks can be used to access methods or fields of an object that
+may or may not be null. If the object is null, then the expression is 
+evaluated as null. If the object is not null, then the expression is evaluated
+as normal.
+```
+var name: String? = null
+var length: Int? = name?.length()
+// length is null
+
+name = "Jane Doe"
+length = name?.length()
+// length is eight
+```
+
+My favorite null safety feature is the Elvis operator `?:`. This operator has 
+earned such a phenomenal name because if you tilt your head sideways, it kind 
+of looks like an emoticon of Elvis with his rock-star hair! It is the 
+null-coalescing operator available in Kotlin and makes working with nullable
+variables even easier. The Elvis operator allows for you to return an 
+alternative value if a statement is null.
+```
+var name: String? = null
+var length: Int // length is a non-nullable Int
+length = name?.length() ?: 0 
+// length is zero
+
+name = "Jane Doe"
+length = name?.length() ?: 0
+// length is eight
+```
+
+## Data Classes
+Kotlin provides the 
+[data class](https://kotlinlang.org/docs/data-classes.html) syntax to make 
+declaring POJOs (Plain Old Java Objects) a one-line affair.
+```
+data class Person(var firstName: String, var lastName: String, var age: Int)
+```
+The data class syntax creates a new class with a default constructor, getters 
+and setters, and commonly used methods such as equals(), toString(), 
+hashCode(), and copy(). All of these defaults can be overridden if necessary,
+but are completely fine for most use cases.
+
+## Class Extensions
+One of the coolest Kotlin features is the ability to extend existing classes by
+defining new methods for that class outside of the class definition. These 
+[extension functions](https://kotlinlang.org/docs/extensions.html) are 
+extremely useful when working with classes that are imported from a library or 
+if you want to install some new logic onto a built-in class.
+
+```
+// MutableList is Kotlin's built-in version of ArrayList
+fun MutableList<Int>.swap(index1: Int, index2: Int) {
+    val tmp = this[index1]
+    this[index1] = this[index2]
+    this[index2] = tmp
+}
+
+val list = mutableListOf(1, 2, 3)
+list.swap(0, 2)
```

Comment:
I assume the list then contains `[3, 2, 1]` - could you add a comment showing the effect as you did for the length stuff above?